File: implementations/typescript-core-20260516/src/digest.ts
import { createHash } from "crypto";
import { blake3 } from "@noble/hashes/blake3.js";

export function digestBytes(data: Buffer, digestAlg: string): Buffer {
  if (digestAlg === "SHA-256") {
    return createHash("sha256").update(data).digest();
  }
  if (digestAlg === "SHA-384") {
    return createHash("sha384").update(data).digest();
  }
  if (digestAlg === "SHA-512") {
    return createHash("sha512").update(data).digest();
  }
  if (digestAlg === "BLAKE3") {
    return Buffer.from(blake3(data));
  }
  if (digestAlg === "BLAKE2b") {
    return createHash("blake2b512").update(data).digest();
  }
  throw new Error(`unknown digest algorithm: ${digestAlg}`);
}

File: cli/src/digest.rs
//! Digest algorithm dispatch for ProofBundle.
//!
//! Supported algorithms: SHA-256, SHA-384, SHA-512, BLAKE3, BLAKE2b-512

use sha2::{Digest as Sha2Digest, Sha256, Sha384, Sha512};

/// Supported digest algorithm identifiers.
pub const DIGEST_ALGORITHMS: &[&str] = &["sha256", "sha384", "sha512", "blake3", "blake2b512"];

#[derive(Debug, Clone, PartialEq, thiserror::Error)]
pub enum DigestError {
    #[error("Unknown digest algorithm: {0}")]
    UnknownAlgorithm(String),
}

/// Compute a digest over the given bytes using the specified algorithm.
pub fn digest_bytes(alg: &str, data: &[u8]) -> Result<Vec<u8>, DigestError> {
    match alg {
        "sha256" => Ok(Sha256::digest(data).to_vec()),
        "sha384" => Ok(Sha384::digest(data).to_vec()),
        "sha512" => Ok(Sha512::digest(data).to_vec()),
        "blake3" => Ok(blake3::hash(data).as_bytes().to_vec()),
        "blake2b512" => {
            let mut hasher = blake2::Blake2b512::new();
            hasher.update(data);
            Ok(hasher.finalize().to_vec())
        }
        other => Err(DigestError::UnknownAlgorithm(other.to_string())),
    }
}

/// Compute a hex-encoded digest string.
pub fn digest_hex(alg: &str, data: &[u8]) -> Result<String, DigestError> {
    let bytes = digest_bytes(alg, data)?;
    Ok(hex::encode(bytes))
}

/// Validate that a digest algorithm identifier is supported.
pub fn validate_digest_alg(alg: &str) -> Result<(), DigestError> {
    if DIGEST_ALGORITHMS.contains(&alg) {
        Ok(())
    } else {
        Err(DigestError::UnknownAlgorithm(alg.to_string()))
    }
}
